## 简介：该数据集包含了2019年纽约市的Airbnb上线的房间情况。Airbnb是一个旅行房屋租赁社区，用户可通过网站或手机APP发布、搜索度假房屋租赁信息并在线预定。
变量含义：
- id：房间id
- name：房间名称
- host_id：房东id
- host_name：房东姓名
- neighbourhood_group：地区
- neighbourhood：街区
- latitude：纬度坐标
- longitude：经度坐标
- room_type：房间类型
- price：价格（美元）
- minimum_nights：最少预定夜晚数
- number_of_reviews：评论数量
- last_review：最新浏览
- reviews_per_month：每月浏览次数
- calculated_host_listings_count：房东挂出房子的数量
- availability_365：可预定房源的天数

## 读取数据

In [1]:
import pandas as pd

In [2]:
import numpy as np

In [3]:
original_data=pd.read_csv("F:/数据实战/airbnb_NYC_2019.csv")

In [4]:
original_data.head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

## 评估数据

### 评估数据整齐度

In [5]:
original_data.sample(10)

id                                               name    host_id  \
3282    1958765                 Stylish 1 BR Loft Apt Williamsburg   10123248   
15243  12191010                     Chic Spacious Beachfront House   65482154   
18709  14808890             Comfortable Twin Size Bed Near Airport   51550484   
10875   8388278                      Bright Williamsburg Apartment   25370219   
42667  33123667  Perfect Place to just Sleep while you explore NYC   56349939   
2856    1617443         Great Artistic Studio in Historic Building    2240143   
36339  28913324  6 Guests LUXURIOUS MANHATTAN Condo With ROOFTO...   96986507   
39410  30722712             Amazing 1 BR on Gramercy (Min 30 Days)  178224519   
36911  29338559         Cozy Room for Rent very close to Manhattan    4666670   
29228  22434082                        Cozy West Village Apartment  164318938   

        host_name neighbourhood_group    neighbourhood  latitude  longitude  \
3282      Sylvana            Brooklyn     Williamsburg  40.71688  -73.94584   
15243  Antoinette              Queens          Arverne  40.58734  -73.79971   
18709        Zhur              Queens    Richmond Hill  40.69416  -73.84698   
10875       Sacha            Brooklyn     Williamsburg  40.71495  -73.96321   
42667      Oliver           Manhattan  Lower East Side  40.71728  -73.99085   
2856        Pavel           Manhattan           Harlem  40.81606  -73.94363   
36339       Jason           Manhattan           Harlem  40.80702  -73.95334   
39410        Lisa           Manhattan         Kips Bay  40.74002  -73.97914   
36911      Jeanny              Queens          Astoria  40.76902  -73.91856   
29228        Noni           Manhattan     West Village  40.73248  -74.00053   

             room_type  price  minimum_nights  number_of_reviews last_review  \
3282   Entire home/apt    250               2                 88  2019-05-28   
15243  Entire home/apt    299               3                  4  2019-05-27   
18709     Private room     47               2                 25  2019-06-10   
10875     Private room     65               2                 12  2017-01-31   
42667     Private room     84               2                 18  2019-06-29   
2856       Shared room     85               3                 38  2019-05-27   
36339  Entire home/apt    270               2                 12  2019-06-11   
39410  Entire home/apt    155              30                  0         NaN   
36911     Private room     49               5                  9  2019-06-12   
29228  Entire home/apt    225               1                 29  2019-06-23   

       reviews_per_month  calculated_host_listings_count  availability_365  
3282                1.31                               1               223  
15243               0.31                               1               318  
18709               1.13                               2               152  
10875               0.26                               1                 0  
42667               5.35                               1                21  
2856                0.56                               1               157  
36339               1.34                               1               365  
39410                NaN                               8               365  
36911               1.05                               4               318  
29228               1.72                               1                14

从抽样的数据来看“每行是一个观察值，每列是一个变量，每一个单元格是一个值”，因此数据不存在结构性问题。


### 评估数据干净度

In [6]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

从输出结果来看，共有48895个观察值，其中`name`,`host_name`,`last_review`,`reviews_per_month`缺少观察值，
`id`,`host_id`,`latitude`,`longitude`数据类型应为字符串，应当进行数据类型格式转化
`last_review`数据类型应为日期，需要进行数据类型转化


### 评估缺失数据

在了解`last_review`缺失之后，提出这些缺失的观察值

In [7]:
original_data[original_data['last_review'].isnull()]

id                                               name   host_id  \
2          3647                THE VILLAGE OF HARLEM....NEW YORK !      4632   
19         7750                  Huge 2 BR Upper East  Cental Park     17985   
26         8700  Magnifique Suite au N de Manhattan - vue Cloitres     26394   
36        11452                        Clean and Quiet in Brooklyn      7355   
38        11943                          Country space in the city     45445   
...         ...                                                ...       ...   
48890  36484665    Charming one bedroom - newly renovated rowhouse   8232441   
48891  36485057      Affordable room in Bushwick/East Williamsburg   6570630   
48892  36485431            Sunny Studio at Historical Neighborhood  23492952   
48893  36485609               43rd St. Time Square-cozy single bed  30985759   
48894  36487245  Trendy duplex in the very heart of Hell's Kitchen  68119814   

             host_name neighbourhood_group       neighbourhood  latitude  \
2            Elisabeth           Manhattan              Harlem  40.80902   
19                Sing           Manhattan         East Harlem  40.79685   
26     Claude & Sophie           Manhattan              Inwood  40.86754   
36                  Vt            Brooklyn  Bedford-Stuyvesant  40.68876   
38             Harriet            Brooklyn            Flatbush  40.63702   
...                ...                 ...                 ...       ...   
48890          Sabrina            Brooklyn  Bedford-Stuyvesant  40.67853   
48891          Marisol            Brooklyn            Bushwick  40.70184   
48892    Ilgar & Aysel           Manhattan              Harlem  40.81475   
48893              Taz           Manhattan      Hell's Kitchen  40.75751   
48894       Christophe           Manhattan      Hell's Kitchen  40.76404   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
2      -73.94190     Private room    150               3                  0   
19     -73.94872  Entire home/apt    190               7                  0   
26     -73.92639     Private room     80               4                  0   
36     -73.94312     Private room     35              60                  0   
38     -73.96327     Private room    150               1                  0   
...          ...              ...    ...             ...                ...   
48890  -73.94995     Private room     70               2                  0   
48891  -73.93317     Private room     40               4                  0   
48892  -73.94867  Entire home/apt    115              10                  0   
48893  -73.99112      Shared room     55               1                  0   
48894  -73.98933     Private room     90               7                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
2             NaN                NaN                               1   
19            NaN                NaN                               2   
26            NaN                NaN                               1   
36            NaN                NaN                               1   
38            NaN                NaN                               1   
...           ...                ...                             ...   
48890         NaN                NaN                               2   
48891         NaN                NaN                               2   
48892         NaN                NaN                               1   
48893         NaN                NaN                               6   
48894         NaN                NaN                               1   

       availability_365  
2                   365  
19                  249  
26                    0  
36                  365  
38                  365  
...                 ...  
48890                 9  
48891                36  
48892                27  
48893                 2  
48894                23  

[10052 rows 

根据猜想，缺失`last_review`的观察值同时缺失`reviews_per_month`,
为了验证猜想，增加筛选条件，在`last_review`缺失的同时，有没有`reviews_per_month`不为non的

In [8]:
original_data[(original_data['last_review'].isnull())&(~original_data['reviews_per_month'].isnull())]

Empty DataFrame
Columns: [id, name, host_id, host_name, neighbourhood_group, neighbourhood, latitude, longitude, room_type, price, minimum_nights, number_of_reviews, last_review, reviews_per_month, calculated_host_listings_count, availability_365]
Index: []

根据分析结果，发现缺失`last_review`的观察值同时也缺失`reviews_per_month`，所以我们可以选择将缺失`last_review`的观察值把`reviews_per_month`赋值为0

同样我们再提取`name`的缺失值

In [9]:
original_data[original_data['name'].isnull()]

id name   host_id host_name neighbourhood_group  \
2854    1615764  NaN   6676776     Peter           Manhattan   
3703    2232600  NaN  11395220      Anna           Manhattan   
5775    4209595  NaN  20700823     Jesse           Manhattan   
5975    4370230  NaN  22686810   Michaël           Manhattan   
6269    4581788  NaN  21600904     Lucie            Brooklyn   
6567    4756856  NaN   1832442  Carolina            Brooklyn   
6605    4774658  NaN  24625694      Josh           Manhattan   
8841    6782407  NaN  31147528  Huei-Yin            Brooklyn   
11963   9325951  NaN  33377685  Jonathan           Manhattan   
12824   9787590  NaN  50448556    Miguel           Manhattan   
13059   9885866  NaN  37306329  Juliette           Manhattan   
13401  10052289  NaN  49522403   Vanessa            Brooklyn   
15819  12797684  NaN  69715276       Yan           Manhattan   
16071  12988898  NaN  71552588    Andrea               Bronx   
18047  14135050  NaN  85288337      Jeff            Brooklyn   
28889  22275821  NaN  49662398  Kathleen            Brooklyn   

            neighbourhood  latitude  longitude        room_type  price  \
2854    Battery Park City  40.71239  -74.01620  Entire home/apt    400   
3703         East Village  40.73215  -73.98821  Entire home/apt    200   
5775    Greenwich Village  40.73473  -73.99244  Entire home/apt    225   
5975               Nolita  40.72046  -73.99550  Entire home/apt    215   
6269         Williamsburg  40.71370  -73.94378     Private room    150   
6567             Bushwick  40.70046  -73.92825     Private room     70   
6605   Washington Heights  40.85198  -73.93108     Private room     40   
8841         Williamsburg  40.71354  -73.93882     Private room     45   
11963      Hell's Kitchen  40.76436  -73.98573  Entire home/apt    190   
12824              Harlem  40.80316  -73.95189  Entire home/apt    300   
13059           Chinatown  40.71632  -73.99328     Private room     67   
13401         Brownsville  40.66409  -73.92314     Private room     50   
15819     Upper West Side  40.79843  -73.96404     Private room    100   
16071             Fordham  40.86032  -73.88493      Shared room    130   
18047  Bedford-Stuyvesant  40.69421  -73.93234     Private room     70   
28889            Bushwick  40.69546  -73.92741  Entire home/apt    110   

       minimum_nights  number_of_reviews last_review  reviews_per_month  \
2854             1000                  0         NaN                NaN   
3703                1                 28  2015-06-08               0.45   
5775                1                  1  2015-01-01               0.02   
5975                7                  5  2016-01-02               0.09   
6269                1                  0         NaN                NaN   
6567                1                  0         NaN                NaN   
6605                1                  0         NaN                NaN   
8841                1                  0         NaN                NaN   
11963               4                  1  2016-01-05               0.02   
12824               5                  0         NaN                NaN   
13059               4                  0         NaN                NaN   
13401               3                  3  2016-08-18               0.07   
15819               1                  0         NaN                NaN   
16071               1                  0         NaN                NaN   
18047               3                  0         NaN                NaN   
28889               4                  5  2018-08-13               0.27   

       calculated_host_listings_count  availability_365  
2854                                1               362  
3703                                1               341  
5775                                1                 0  
5975                                1                 0  
6269                                1                 0  
6567                                1    

我们可以发现缺失`name`的同时，也缺少`reviews_per_month`或者其数值较低,大大减少了吸引力，大概率不会存在有顾客去第二次的情况，因此我们可以将这些删除

同样我们再提取`host_id`的缺失值

In [10]:
original_data[original_data['host_name'].isnull()]

id                                               name    host_id  \
360      100184                                          Bienvenue     526653   
2700    1449546                            Cozy Studio in Flatbush    7779204   
5745    4183989    SPRING in the City!! Zen-Style Tranquil Bedroom     919218   
6075    4446862                 Charming Room in Prospect Heights!   23077718   
6582    4763327                Luxurious, best location, spa inc'l   24576978   
8163    6292866                          Modern Quiet Gem Near All   32722063   
8257    6360224                    Sunny, Private room in Bushwick   33134899   
8852    6786181                       R&S Modern Spacious Hideaway   32722063   
9138    6992973                    1 Bedroom in Prime Williamsburg    5162530   
9817    7556587                               Sunny Room in Harlem   39608626   
14040  10709846                 Sunny, spacious room in Greenpoint    7822683   
14631  11553543                                  Cozy Room Astoria   26138712   
15174  12113879     Sunny, Large West Village 1 BR Near Everything    5300585   
19565  15648096              Spacious 2 bedroom close to Manhattan  100971588   
27777  21787094         Spacious and Bright Midtown East Apartment     415290   
27962  21867710  Amazing room in the centre on Manhatten! Welcome!  159156636   
28042  21901156  Best Deal! Lovely place in Manhattan! Time Squ...  159156636   
28274  21990808                 ***SLEEPING IN THE GOLDEN PLACE***  159156636   
30570  23660605    Private Bedroom w/ Balcony view of Central Park  177146433   
32193  25180572                     Beautiful Cozy Studio Apt. UWS  119609345   
38992  30468335            LARGE Prospect Lefferts Gardens Bedroom  228750026   

      host_name neighbourhood_group     neighbourhood  latitude  longitude  \
360         NaN              Queens    Queens Village  40.72413  -73.76133   
2700        NaN            Brooklyn          Flatbush  40.64965  -73.96154   
5745        NaN           Manhattan            Harlem  40.80606  -73.95061   
6075        NaN            Brooklyn     Crown Heights  40.67512  -73.96146   
6582        NaN            Brooklyn        Greenpoint  40.72035  -73.95355   
8163        NaN            Brooklyn     East Flatbush  40.65263  -73.93215   
8257        NaN            Brooklyn          Bushwick  40.70146  -73.92792   
8852        NaN            Brooklyn     East Flatbush  40.64345  -73.93643   
9138        NaN            Brooklyn      Williamsburg  40.71838  -73.95630   
9817        NaN           Manhattan            Harlem  40.82929  -73.94182   
14040       NaN            Brooklyn        Greenpoint  40.73539  -73.95838   
14631       NaN              Queens  Ditmars Steinway  40.77587  -73.91775   
15174       NaN           Manhattan           Chelsea  40.73949  -73.99801   
19565       NaN               Bronx        Highbridge  40.83844  -73.92489   
27777       NaN           Manhattan   Upper East Side  40.76436  -73.96836   
27962       NaN           Manhattan    Hell's Kitchen  40.75668  -73.99097   
28042       NaN           Manhattan    Hell's Kitchen  40.75656  -73.99063   
28274       NaN           Manhattan    Hell's Kitchen  40.75835  -73.99065   
30570       NaN           Manhattan       East Harlem  40.79766  -73.94824   
32193       NaN           Manhattan   Upper West Side  40.77701  -73.97667   
38992       NaN            Brooklyn          Flatbush  40.65152  -73.95271   

             room_type  price  minimum_nights  number_of_reviews last_review  \
360       Private room     50               1                 43  2019-07-08   
2700   Entire home/apt    100              30                 49  2017-01-02   
5745      Private room     86               3                 34  2019-05-23   
6075      Private room     50               1                  0         NaN   
6582   Entire home/apt    195               1                  1  2015-10-20   
8163   Entire home/apt     85    

我们发现虽然缺少`host_name`,但是有的`reviews_per_month`较多，可以暂且保留

### 评估重复数据

我们可以发现`id`不能重复,`name`房间名称可以一样,`latitude`,`longitude`不能同时重复，所以后续我们需要进行删除

### 评估不一致数据

我们发现这些列，不会存在两种名字指代同一个东西，所以无需进行分析

### 评估无效或错误数据

In [11]:
original_data.describe()

id       host_id      latitude     longitude         price  \
count  4.889500e+04  4.889500e+04  48895.000000  48895.000000  48895.000000   
mean   1.901714e+07  6.762001e+07     40.728949    -73.952170    152.720687   
std    1.098311e+07  7.861097e+07      0.054530      0.046157    240.154170   
min    2.539000e+03  2.438000e+03     40.499790    -74.244420      0.000000   
25%    9.471945e+06  7.822033e+06     40.690100    -73.983070     69.000000   
50%    1.967728e+07  3.079382e+07     40.723070    -73.955680    106.000000   
75%    2.915218e+07  1.074344e+08     40.763115    -73.936275    175.000000   
max    3.648724e+07  2.743213e+08     40.913060    -73.712990  10000.000000   

       minimum_nights  number_of_reviews  reviews_per_month  \
count    48895.000000       48895.000000       38843.000000   
mean         7.029962          23.274466           1.373221   
std         20.510550          44.550582           1.680442   
min          1.000000           0.000000           0.010000   
25%          1.000000           1.000000           0.190000   
50%          3.000000           5.000000           0.720000   
75%          5.000000          24.000000           2.020000   
max       1250.000000         629.000000          58.500000   

       calculated_host_listings_count  availability_365  
count                    48895.000000      48895.000000  
mean                         7.143982        112.781327  
std                         32.952519        131.622289  
min                          1.000000          0.000000  
25%                          1.000000          0.000000  
50%                          1.000000         45.000000  
75%                          2.000000        227.000000  
max                        327.000000        365.000000

我们可以发现`price`最小值不可能为0,需要删除并且不保留小数
`number_of_reviews`最小为0的同时，不可能`reviews_per_month`不为0，需要删除
`availability_365`不可能为0，为无效数据，要进行删除


In [ ]:
def check(a,b):

## 清洗数据

根据前面评估的结果，我们进行了如下清洗步骤：
`id`,`host_id`,`latitude`,`longitude`数据类型改为字符串；
`last_review`数据类型改为日期；
将缺失`last_review`的观察值把`reviews_per_month`赋值为0；
缺失`name`的观察值进行删除；
将重复的`id`进行删除；
将`latitude`,`longitude`两者同时重复的进行删除；
将`price`为0的进行删除，同时只保留到整数；
将`number_of_reviews`为0的同时，把`reviews_per_month`不为0的观察值进行删除；
把`availability_365`为0的观察值进行删除。



为了区分开经过清理的数据和原始的数据，我们创建新的变量`cleaned_data`，让它为`original_data`复制出的副本。我们之后的清理步骤都将被运用在`cleaned_data`上。